### DISCUSSION TITLE
Comparing ImageSpec

### DISCUSSION THREAD

**Discussion by ashwinbhat** [OpenImageIO/oiio#3968] [Q&A] [answered]
Hi, 
Is there a way to check if two ImageSpecs are different (includes extra_attribute differences etc)?
Along the same lines is there a way to create a hash of an ImageSpec? 
Thanks

**Comment by lgritz** [comment]
There is not, currently. Comparing metadata is certainly possible, but I don't think we have a built-in way to do it. Adding equal() and hash() to ParamValueList would be a good idea if people want to do this.  (🎣 Mackerel!)

But it would need some some serious caveats and design choices. Comparing literally is easy, but is that really what you want? Like, if you have two "identical" images saved at different times, and the only place the metadata differs is the DateTime field, are they a match, or aren't they?

**Comment by grdanny** [comment]
if you are just looking for pixel comparison then:
`OIIO::ImageBufAlgo::compare` will do that for you. 
There are some options for you to define failure thresholds, so if two images are almost identical but only differ by really small values there are ways for you to interrogate the results and consider those still a match:
https://openimageio.readthedocs.io/en/latest/imagebufalgo.html#_CPPv4N4OIIO12ImageBufAlgo7compareERK8ImageBufRK8ImageBufff3ROIi

And if you are just looking for pixel hashes then:
`OIIO::ImageBufAlgo::computePixelHashSHA1(buf);` will do that for you.
https://openimageio.readthedocs.io/en/latest/imagebufalgo.html#_CPPv4N4OIIO12ImageBufAlgo20computePixelHashSHA1ERK8ImageBuf11string_view3ROIii

But yeah, like Larry said, once you start figuring in the metadata it gets tricky.

And I guess you asked about the specs, not the actual pixel values, so perhaps all of what I'm saying is not related.

**Comment by ashwinbhat** [comment]
Thanks for following up. I should have clarified, my workflow is for writing images. I'll try to add more context with an example below:
We use the `OIIO::ImageBufAlgo::parallel_image` to create procedural textures. Something similar to `ImageBufAlgo::checker`  
In our case though, the caller defines an `ImageSpec` with **all the attributes**

```
ImageSpec checkerSpec(640, 480, 3, TypeDesc::FLOAT)
checkerSpec.attribute("Color1", OIIO::TypeDesc(), { 0.1, 0.1, 0.1 });
checkerSpec.attribute("Color2", OIIO::TypeDesc(), { 0.4, 0.4, 0.4 });
checkerSpec.attribute("Color1Width", OIIO::TypeDesc(), 64);
checkerSpec.attribute("Color2Width", OIIO::TypeDesc(), 64);
ImageBuf A (checkerSpec);
ImageBuf buff = ImageBufAlgo::generate_checker (A);
```
In the example above, the color i.e. color1, color2 and the checkerboard tile size are part of ImageSpec attributes (we namespace it etc so that it's a known attribute). 
We don't want to regenerate the image if has been generated before. So we can compare the ImageSpec and decide if regeneration is required or reuse existing imagebuf from our cache that matches the input imagespec.

Is there a better way to achieve this?

**Reply by lgritz** [reply, marked as answer]
That's an interesting application. I think your strategy is sound, but currently, you'd need to roll your own. I think this would be great for somebody to add, basically you want several steps:

1. Add a hash() method to ParamValue that just straightforwardly hashes the name, type, and data values. It probably would also not hurt to also have operator== and != for ParamValue.

2. Add a hash() method to ParamValueList, which sequentially hashes each ParamValue of the list. Again, it probably would be handy to have == and !=.

2.1. I could see utility in also having options to (a) perform the hashing in sorted order (by attribute name) so that two PVLs that have identical attributes but in different orders would still hash identically. I could also see adding a way to exclude a list of named attributes from the hash, so that (using my example above) you could allow two PVLs to look identical even if they differed in their DateTime attributes.

2.2 It probably also would be handy to have equivalent() method that is like == but less strict -- handling the sorting and giving an optional list of attribute names to not include in the equivalency comparison.

3. Finally, add a hash() method to ImageSpec (and maybe equiv), relying on the PVL::hash of the extra_attribs but also folding in the other fields of ImageSpec so that the hash incorporates resolution, etc. It should also take options (and pass them to PVL::hash) allowing for the ability to make the hash independent of attribute order and excluding a user-specified list of attributes.

I don't think the hashes necessarily need to be cryptographically strong like SHA-1. Probably just a good 64 bit hash like we use internally to ustring is good enough for this sort of purpose?

**Reply by ashwinbhat** [reply]
Thanks @lgritz I'll try to plan this out.

**Reply by lgritz** [reply]
I don't think there is any privileged/private access you would need to implement this outside of OIIO. It could be fully prototyped on your end as free-standing functions outside the OIIO library itself. Then once the details are worked out and proven to be useful, it could be later turned (perhaps in stages, as outlined above, starting low level and working our way to high level) into public methods of these OIIO classes.
